### PR TITLE
159 Add framing streams

### PR DESCRIPTION
**Pull request by Miceuz**
#439 Remove CRC32 implementation 
#440 uart dma update
#441 159 frame detector
👉#442 Add framing streams 👈 **YOU ARE HERE**
#443 frame decoding
#444  Update comms to use UART_DMA and framing

### CODE REVIEW

File: common/libs/framing/framing_streams.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      r += output.Put(FRAMING_MARK);
+      r += output.Put(b);
+      FrameDone();
+      return r;
+      break;
+    case FRAMING_MARK:
+    case FRAMING_ESC:
+      r = StartFrameIfDone();
+      r += output.Put(FRAMING_ESC);
+      r += output.Put(b ^ 0x20);
```
Comment (jkff) [inline common/libs/framing/framing_streams.h:111]: Why ^ 0x20?

File: common/libs/framing/framing_streams.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    };
+  }
+
+private:
+  StreamResponse StartFrameIfDone() {
+    return is_frame_done ? output.Put(FRAMING_MARK)
+                         : (StreamResponse){0, STREAM_SUCCESS};
+  }
+
+  static bool shouldEscape(uint8_t b) {
```
Comment (jkff) [inline common/libs/framing/framing_streams.h:130]: This function is unused

File: common/libs/framing/framing_streams.h [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    return FRAMING_MARK == b || FRAMING_ESC == b;
+  }
+
+  void FrameDone() {
+    if (!is_frame_done) {
```
Comment (jkff) [inline common/libs/framing/framing_streams.h:135]: This function is equivalent to simply saying `is_frame_done = true;` and the next one to `is_frame_done = false;` so they both can be inlined.

File: common/libs/framing/framing_streams.h [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    }
+  }
+  void FrameStartedMaybe() {
+    if (is_frame_done) {
+      is_frame_done = false;
+    }
+  }
+};
+
+// Just returns number of bytes written, designed to be used for evaluation of
```
Comment (jkff) [inline common/libs/framing/framing_streams.h:146]: This doesn't seem to be used

File: common/libs/framing/framing_streams.h [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+  StreamResponse Put(int32_t b) {
+    StreamResponse r = {0, 0};
+    switch (b) {
+    case END_OF_STREAM:
+      r = StartFrameIfDone();
+      r += output.Put(FRAMING_MARK);
+      r += output.Put(b);
+      FrameDone();
+      return r;
+      break;
```
Comment (jkff) [inline common/libs/framing/framing_streams.h:106]: break after return is unnecessary

File: common/libs/framing/framing_streams.h [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      r = StartFrameIfDone();
+      r += output.Put(b);
+      FrameStartedMaybe();
+      return r;
+      break;
+    };
+  }
+
+private:
+  StreamResponse StartFrameIfDone() {
```
Comment (jkff) [inline common/libs/framing/framing_streams.h:125]: This function is always called as "r = StartFrameIfDone()" which is equivalent to `if (is_frame_done) output.Put(FRAMING_MARK);` and it seems easier to inline this.

File: common/libs/framing/framing_streams.h [hunk 7]
```diff
@@ -0,0 +1,159 @@
+#ifndef __FRAMING_STREAMS_H_
+#define __FRAMING_STREAMS_H_
+
+#include "checksum.h"
+#include "framing_spec_chars.h"
+#include <stdint.h>
+
+constexpr int32_t END_OF_STREAM = -1;
+
+enum ResponseFlags {
```
[the remark this answers is not in the document]
Comment (Miceuz) [reply common/libs/framing/framing_streams.h:10]: Errr, no? 
We want to binary OR the flags which will result in a combination that is not in the enum.

File: common/libs/framing/framing_spec_chars.h [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#ifndef __FRAMING_SPEC_CHARS_H__
```
[the remark this answers is not in the document]
Comment (Miceuz) [reply common/libs/framing/framing_spec_chars.h:1]: DONE

File: common/libs/framing/framing_streams.h (hunk 5 above)
Comment (Miceuz) [reply common/libs/framing/framing_streams.h:106]: doh. FIXED

File: common/libs/framing/framing_streams.h (hunk 6 above)
Comment (Miceuz) [reply common/libs/framing/framing_streams.h:122]: I disagree, code looks much cleaner with the function refactored. It was like you are suggesting before - a lot of code duplication and noise.

**Pull request closed without merging**